contract/cosmwasm/compile.go

```diff
@@ -0,0 +1,119 @@
+package cosmwasm
+
+import (
+	"context"
+	"path/filepath"
+	"fmt"
+	"runtime"
+	"io"
+	"os"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/errdefs"
+	"github.com/docker/docker/pkg/stdcopy"
+)
+
+// compile will compile the specified repo using the specified docker image and version
+func compile(image string, version string, repoPath string) (string, error) {
+	// Set the image to pull/use
+	arch := ""
+	if runtime.GOARCH == "arm64" {
+		arch = "-arm64"
+	}
+	imageFull := image + arch + ":" + version
+
+	// Get absolute path of contract project
+	pwd, err := os.Getwd()
+	if err != nil {
+		return "", fmt.Errorf("getwd: %w", err)
+	}
+	repoPathFull := filepath.Join(pwd, repoPath)
+
+	ctx := context.Background()
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return "", fmt.Errorf("new client with opts: %w", err)
+	}
+	defer cli.Close()
+
+	reader, err := cli.ImagePull(ctx, imageFull, types.ImagePullOptions{})
+	if err != nil {
+		return "", fmt.Errorf("pull image %s: %w", imageFull, err)
+	}
+
+	defer reader.Close()
+	_, err = io.Copy(os.Stdout, reader)
+	if err != nil {
+		return "", fmt.Errorf("io copy %s: %w", imageFull, err)
+	}
+
+	resp, err := cli.ContainerCreate(ctx, &container.Config{
+		Image: imageFull,
+		Tty:   false,
+	}, &container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type: mount.TypeBind,
+				Source: repoPathFull,
+				Target: "/code",
+			},
+			{
+				Type: mount.TypeVolume,
+				Source: filepath.Base(repoPathFull)+"_cache",
+				Target: "/target",
```

Comment:
Just a note: this will break contracts using rust-optimizer v0.12.13 and below *(target moved from /code/target to /target in v13)*.

I mention this as I see many still building with the v0.12 suite in CI, including myself just last week.

Reply:
Thanks, updated!